gusto/state.py

```diff
@@ -267,16 +267,22 @@ def setup_diagnostics(self):
 
     def setup_dump(self, pickup=False):
 
-        # setup dump files
         # check for existence of directory so as not to overwrite
         # output files
+        # setup dump files
+        # setup checkpoint files
         self.dumpdir = path.join("results", self.output.dirname)
         outfile = path.join(self.dumpdir, "field_output.pvd")
         if self.mesh.comm.rank == 0 and "pytest" not in self.output.dirname \
            and path.exists(self.dumpdir) and not pickup:
             raise IOError("results directory '%s' already exists" % self.dumpdir)
         self.dumpcount = itertools.count()
         self.dumpfile = File(outfile, project_output=self.output.project_fields, comm=self.mesh.comm)
+        self.chkpts = ["chkpt", "chkptbk"]
+        for name in self.chkpts:
```

Comment:
I'm confused about the two different checkpoint files.  It just looks you write the same data twice.  I know that's what was happening previously, but perhaps you can explain.

Reply:
As far as I remember this was done for the case if something goes wrong and the test run stops while writing to the checkpoint. If the checkpoint doesn't work one would then still have the other checkpoint to recover the data.

Reply:
OK, this is a really bad way to do file rotation, FWIW.  One should do something like:

```
os.rename("current", "backup")
dump_checkpoint("current")
```

Reply:
I'm not sure how to use os.rename without having to build a new checkpoint each time (leading to the original error). How about if I build one checkpoint in setup_dump, then just copy the file each time in dump after having written to it, using something like
''''
from shutil import copyfile
copyfile(path.join(self.dumpdir,"chkpt.h5"), path.join(self.dumpdir,"chkptbk.h5"))
''''
I can't really think of a scenario where this wouldn't produce at least one functioning backup file while the version of writing to two files in a row (as it is now) would.

Reply:
So if I understand the problem, what you're trying to avoid is the simulation crashing overwriting some "good" checkpoint file.  So think about the time loop:

At the beginning you have no checkpoint files.
After step one, you either have one checkpoint file (`current`) or none (if the simulation crashed).

Now you come to step two and want to do a checkpoint.  So you first move `current` to `backup`, ensuring that if you now crash there is a backup checkpoint file from the previous dump.

Now you write to `current`.  If this succeeds, great, if not, then you still have the backup file.

This is similar to your suggestion, but doesn't require as much filesystem movement (`os.rename` doesn't move data, just metadata).

So the logic is like:

```python
while t < tend:
    if dump:
        if os.path.exists("checkpoint.h5"):
            os.rename("checkpoint.h5", "backup.h5")
        dump_checkpoint("checkpoint.h5")
```

Reply:
Thanks for the clarification! As I understand it, I will run into the following issue with this:
Suppose I have one checkpoint file (```current```) after step one. Then os.rename moves the metadata, but not the data, renaming the file to backup.h5.

But then with ```dump_checkpoint``` I will have to create a new data set, since I do not want to write to the existing data set (labelled as backup.h5). Hence, I create a new data set, naming it "checkpoint.h5", which is then renamed using ```os.rename``` in the next time step and so forth.

This means, if I understand this right, that if I want to use ```os.rename``` I will have to create a new data set using ```dump_checkpoint``` each time, which is what I want to avoid in the first place (since ```PETSc.ViewerHDF5().create``` in  ```Dumbcheckpoint.new_file()``` produces an error after 240 or so dump steps).

This is why I thought I should use ```copyfile``` instead, since this way I can avoid using ```dump_checkpoint``` each time.

Out of interest (this only makes sense if I understood this right): when using ```copyfile```, I will only ever have two data sets, since the backup one is replaced each time ```copyfile``` is used (according to their documentation). What happens in the ```os.rename``` scenario? I don't really see any data sets being deleted or replaced in this case.

Reply:
OK, I see.  Yes, what you're proposing makes sense I think.  So you always write to `checkpoint.h5` and intermittently copy it to `backup.h5`.